### ISSUE TITLE
[FR] File chooser dialog example

### ISSUE THREAD

**Issue by ITwrx** [StefanSalewski/gintro#101]
It would be nice if there were a gtk file chooser dialog example for gintro, like this: https://python-gtk-3-tutorial.readthedocs.io/en/latest/dialogs.html#filechooserdialog minus the filters function.  

I have started, but have run into some issues i'm not sure how to solve, and instead of asking in a forum about each one, i hope an example would teach me/others how to find that type of information in the future. I have looked through the existing gintro examples and gtk.nim and i'm still stuck.

I'm hoping to build many things using gintro and nim, so thanks for all the hours you put in already. 

Also, it would be nice if you added a freedom and privacy-respecting method of donation to your personal site. like an xmr address, for instance.

thanks

**Comment by StefanSalewski** [comment]
Yes, a file chooser dialog should be not that hard. How urgent do you need it?

I am currently mostly working on the more difficult parts for the GTK4 book:

http://ssalewski.de/gtkprogramming.html

Tried to describe the general behaviour of GtkApplication class with all the startup code, and now working on explaining GAction. There are already GAction examples in the gintro README, but the explanations are not yet good enough.

Donations: Makes not really sense currently, the number of Nim GTK users is too small still. Maybe 10 really active users? So no chance for a monetary compensation for 1500 hours work on gintro, and 600 hours on oldgtk3. But maybe in a few years we will have at least a few dozen users?

**Comment by ITwrx** [comment]
I could use it ASAP, as the little program i am trying to write requires it, but i don't want to be demanding. Also, i think i may have found the info for adding the buttons i was looking for, so you might be able to wait me out. :)

I came across your gtk book yesterday, and posted a link on the fediverse. It already answered one question i had previously.

Regarding donations, it's up to you, of course, and i don't mean to imply that i could send you a lamborghini down payment (more like beverage money, every once in a while until i have a funded/successful project using gintro), but if you don't have a means of donation, people will think you're independently wealthy and not wanting/needing compensation.

**Comment by ITwrx** [comment]
also, i thought if i followed an example i might learn more about how to use gintro properly instead of whatever hack i cobble together, for now. :)

**Comment by StefanSalewski** [comment]
Well, I just remembered that there is already a very basic example available:

https://github.com/StefanSalewski/gintro/issues/27

See bottom of page, seems still to compile and work. Maybe that is enough to get you starting.

Unfortunately that guy closed the issue and vanished, so I forgot.

I will try to provide a really nice example, but that is generally some work, I would have to read the C API docs and explain all well, so that is generally a full weekend. And there seems to be a new native dialog, I have still to learn it. In gtk4/tests there is a C example available, but it is about 500 lines, so also some work to understand and convert to Nim.

Let me know if you still need more support.

**Comment by ITwrx** [comment]
i hadn't seen that. That's all i needed. My dialog is working now. Thanks a lot! 

One thing i don't understand is what is the `ord` in `discard dialog.addButton("OK", ResponseType.accept.ord)` and where is that `ord` coming from, so i can know where i was supposed to have looked.

Also, you might consider adding that example to the examples directory, as it's pretty useful how it is.

**Comment by StefanSalewski** [comment]
Ord is short for ordinal, it works on enums and some other data types and returns the corresponding integer value. In most cases int() seems to work also, but in manual ord is recommended. I think the reason why ord was necessary was that responses like cancel are enums, but that the function returns int. The general problem with enums is that enums can not be expanded, we declare some enums values in a module, and other modules or other users can not expand it. But responses generally can expand, there can be a "maybe" or "when sun shines" as response.

And yes, I should add as filechooser example to the readme or the gtk4 book. Maybe I had done that in the past already, but the one who asked for it closed the issue, so I forgot about it.
